### DISCUSSION TITLE
Updating state on menu event

### DISCUSSION THREAD

**Discussion by FrankLaterza** [tauri-apps/tauri#7707] [Q&A] [answered]
My goal is to share data with the state manager and update it through the menu events. What would be the best approach for this? Currently, I have the state manager set up in the tauri builder and I'm trying to update the state in the menu event. I thought I found a solution by getting the state from app handle, but every time I run the "open" menu event, the program crashes.

On "open" menu event: ```thread 'main' panicked at 'state() called before manage() for given type'```

```rust
// data struct
pub struct Data {
    file_path: Option<PathBuf>,
}

pub struct AppData(Mutex<Data>);

fn main() {
    // tauri builder
    tauri::Builder::default()
        .manage(AppData(
            // create a temp path
            Mutex::new(Data {
                file_path: Some(PathBuf::from("/dir")),
            }),
        ))
        .invoke_handler(tauri::generate_handler![])
        .menu(
            Menu::new()
                //.add_item(CustomMenuItem::new("file", "File"))
                .add_submenu(Submenu::new(
                    "File",
                    Menu::new().add_item(CustomMenuItem::new("open", "Open")),
                ))
        )
        // try to update the managed state by getting app handle from event. 
        .on_menu_event(|event| match event.menu_item_id() {
            "open" => {
                let app = event.window().app_handle();
                let state = app.state::<State<AppData>>();
                let mut guard = state.0.lock().unwrap();
                guard.file_path = Some(PathBuf::from("/dir"));
            }
            _ => {}
        })
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}
```

**Comment by FabianLars** [comment, marked as answer]
I think it should be as simple as changing this `let state = app.state::<State<AppData>>();` to this `let state = app.state::<AppData>();`. The state() function only wants to know your own custom type and it uses that to find the state in the store so if you wrap it with `State` (even tho it comes out wrapped in `State`) it looks for a different state.

**Reply by FrankLaterza** [reply]
My mistake. Thank you for the correction!
